### PR TITLE
feat(audit-service): export feature added

### LINKED ISSUE #1449: @sourceloop/audit-service: Ability to Export Audit logs
**Is your feature request related to a problem? Please describe.**
Need the micro service to be able to export logs 

**Describe the solution you'd like**
Data shall be fetched from S3(archived logs) and primary database which will be stored in an excel file(by default) at a location as provided by the user. The file can also have custom columns

### PR DESCRIPTION
**Pull request by arpit1503khanna**
## Description
export feature added

Fixes #1449

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [x] Performed a self-review of my own code
- [x] npm test passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [x] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/audit-service/src/component.ts [hunk 1]
```diff
@@ -87,12 +99,20 @@ export class AuditServiceComponent implements Component {
 
     this.models = [AuditLog, MappingLog, Job];
 
-    this.controllers = [AuditController, ArchiveLogController];
+    this.controllers = [
+      AuditController,
+      ArchiveLogController,
+      ExportLogsController,
```
Comment (shubhamp-sf) [inline services/audit-service/src/component.ts:105]: Merge this into audit controller.

File: services/audit-service/src/component.ts [hunk 2]
```diff
@@ -87,12 +99,20 @@ export class AuditServiceComponent implements Component {
...
+    ];
 
     this.providers[QuerySelectedFilesServiceBindings.QUERY_ARCHIVED_LOGS.key] =
       QuerySelectedFilesProvider;
     this.providers[ExportToCsvServiceBindings.EXPORT_LOGS.key] =
       ExportToCsvProvider;
+    this.providers[AuditLogExportServiceBindings.EXPORT_AUDIT_LOGS.key] =
+      AuditLogExportProvider;
+    this.providers[ColumnBuilderServiceBindings.BUILD_COLUMN.key] =
```
Comment (shubhamp-sf) [inline services/audit-service/src/component.ts:114]:
```suggestion
    this.providers[ColumnBuilderServiceBindings.COLUMN_BUILDER.key] =
```

File: services/audit-service/src/controllers/audit.controller.ts [hunk 3]
```diff
@@ -142,6 +144,7 @@ export class AuditController {
       const job = (await this.jobRepository.create({
         filterUsed: filter,
         status: FileStatusKey.Pending,
+        operation: OperationKey.Query,
       })) as Job;
```
Comment (shubhamp-sf) [inline services/audit-service/src/controllers/audit.controller.ts:148]: Why `as Job`?

File: services/audit-service/src/controllers/export-logs.controller.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  async find(
+    @param.query.boolean('includeArchivedLogs')
+    includeArchivedLogs: boolean,
+    @param.filter(AuditLog) filter?: Filter<AuditLog>,
+  ): Promise<object> {
+    if (includeArchivedLogs) {
+      const job = (await this.jobRepository.create({
+        filterUsed: filter,
+        status: FileStatusKey.Pending,
+        operation: 'EXPORT',
```
Comment (shubhamp-sf) [inline services/audit-service/src/controllers/export-logs.controller.ts:65]: Change this to enum.

File: services/audit-service/src/controllers/export-logs.controller.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      })) as Job;
```
Comment (shubhamp-sf) [inline services/audit-service/src/controllers/export-logs.controller.ts:66]:
```suggestion
      }));
```

File: services/audit-service/src/controllers/export-logs.controller.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            schema: {
+              type: 'array',
+              items: getModelSchemaRef(AuditLog, {includeRelations: true}),
+            },
+          },
+        },
+      },
+    },
+  })
+  async find(
```
Comment (shubhamp-sf) [inline services/audit-service/src/controllers/export-logs.controller.ts:56]:
```suggestion
  async export(
```

File: services/audit-service/src/controllers/export-logs.controller.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    @param.query.boolean('includeArchivedLogs')
+    includeArchivedLogs: boolean,
+    @param.filter(AuditLog) filter?: Filter<AuditLog>,
+  ): Promise<object> {
```
Comment (shubhamp-sf) [inline services/audit-service/src/controllers/export-logs.controller.ts:60]: be more specific.

File: services/audit-service/src/enums/operation-key.enum.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export const enum OperationKey {
+  Export = 'EXPORT',
+  Query = 'QUERY',
```
Comment (shubhamp-sf) [inline services/audit-service/src/enums/operation-key.enum.ts:3]: Capitalize the keys

File: services/audit-service/src/keys.ts [hunk 9]
```diff
@@ -28,3 +30,13 @@ export namespace ExportToCsvServiceBindings {
     `${BINDING_PREFIX}.audit.archiveLogs`,
   );
 }
+export namespace AuditLogExportServiceBindings {
+  export const EXPORT_AUDIT_LOGS = BindingKey.create<AuditLogExportFn>(
+    `${BINDING_PREFIX}.audit.exportAuditLogs`,
+  );
+}
+export namespace ColumnBuilderServiceBindings {
+  export const BUILD_COLUMN = BindingKey.create<ColumnBuilderFn>(
```
Comment (shubhamp-sf) [inline services/audit-service/src/keys.ts:39]: reflect the name change here too

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+@injectable({scope: BindingScope.TRANSIENT})
+export class AuditLogExportProvider implements Provider<AuditLogExportFn> {
+  constructor(
+    @inject(LOGGER.LOGGER_INJECT)
+    public logger: ILogger,
+  ) {}
+
+  value(): AuditLogExportFn {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    return async (data: any[]) => {
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:18]: Don't use any

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    };
+  }
+  async auditLogExport(data: AnyObject[]) {
+    const targetDirectoryPath = process.env.PATH_TO_EXPORT_FILES_FOLDER;
+    if (!targetDirectoryPath) {
+      throw new Error('Path to target directory not defined');
+    }
+    data = data.map(e => e.toJSON());
+
+    const targetDirectory = path.join(targetDirectoryPath, '/Export-logs/');
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:29]:
```suggestion
    const targetDirectory = path.join(targetDirectoryPath, '/export-logs/');
```

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    fs.mkdir(targetDirectory, error =>
+      error
+        ? this.logger.error('Error in Export Logs Directory')
+        : this.logger.info('Target Directory Created!'),
+    );
+    const fileName = `auditLog_records${new Date().toISOString()}.xlsx`;
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:35]:
```suggestion
    const fileName = `auditlog-records${new Date().toISOString()}.xlsx`;
```

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    const filepath = path.join(targetDirectory, fileName);
+    const worksheetName = 'auditlogs';
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:37]:
```suggestion
    const worksheetName = 'Audit Logs';
```

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    const worksheetColumnNames = Object.keys(data[0]);
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:38]: Handle empty data cases.

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+  }
+  async auditLogExport(data: AnyObject[]) {
+    const targetDirectoryPath = process.env.PATH_TO_EXPORT_FILES_FOLDER;
+    if (!targetDirectoryPath) {
+      throw new Error('Path to target directory not defined');
+    }
+    data = data.map(e => e.toJSON());
+
+    const targetDirectory = path.join(targetDirectoryPath, '/Export-logs/');
+    fs.mkdir(targetDirectory, error =>
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:30]: wait for it be completed.

File: services/audit-service/src/services/column-builder.service.ts [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+export class ColumnBuilderProvider implements Provider<ColumnBuilderFn> {
+  constructor() {}
+
+  value(): ColumnBuilderFn {
+    return async (auditLogs: AuditLog[]) => {
+      return this.columnBuilder(auditLogs);
+    };
+  }
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async columnBuilder(auditLogs: AuditLog[]): Promise<any[]> {
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/column-builder.service.ts:15]: Don't use any.

File: services/audit-service/src/types.ts [hunk 17]
```diff
@@ -20,7 +20,10 @@ export type QuerySelectedFilesFn = (
 export type ExportToCsvFn = (
   selectedAuditLogs: AuditLog[],
 ) => Promise<AWS.S3.ManagedUpload.SendData>;
-
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type AuditLogExportFn = (data: any[]) => Promise<void>;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type ColumnBuilderFn = (auditLogs: AuditLog[]) => Promise<any[]>;
```
Comment (shubhamp-sf) [inline services/audit-service/src/types.ts:26]: Any can be avoided here.

**Review by shubhamp-sf — CHANGES_REQUESTED** [review]

File: services/audit-service/src/services/job-processing.service.ts [hunk 18]
```diff
@@ -73,16 +88,25 @@ export class JobProcessingService {
...
+      }
     } catch (error) {
       const newJob = new Job({
         id: jobId,
-        status: FileStatusKey.Failed,
+        status: FileStatusKey.FAILED,
         result: '',
+        operation: job.operation,
       });
       await this.jobRepository.updateById(jobId, newJob);
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/job-processing.service.ts:111]: Don't we just need to update the status? Why pass all other options as well?

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    }
+    data = data.map(e => e.toJSON());
+
+    const targetDirectory = path.join(targetDirectoryPath, '/export-logs/');
+    fs.mkdir(targetDirectory, error =>
+      error
+        ? this.logger.error('Error in Export Logs Directory')
+        : this.logger.info('Target Directory Created!'),
+    );
+    const fileName = `auditLog-records${new Date().toISOString()}.xlsx`;
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:34]:
```suggestion
    const fileName = `audit-logs-${new Date().toISOString()}.xlsx`;
```

File: services/audit-service/src/keys.ts [hunk 20]
```diff
@@ -28,3 +30,13 @@ export namespace ExportToCsvServiceBindings {
...
   );
 }
+export namespace AuditLogExportServiceBindings {
+  export const EXPORT_AUDIT_LOGS = BindingKey.create<AuditLogExportFn>(
+    `${BINDING_PREFIX}.audit.exportAuditLogs`,
+  );
+}
+export namespace ColumnBuilderServiceBindings {
+  export const COLUMN_BUILDER = BindingKey.create<ColumnBuilderFn>(
+    `${BINDING_PREFIX}.audit.columnBuild`,
```
Comment (shubhamp-sf) [inline services/audit-service/src/keys.ts:40]:
```suggestion
    `${BINDING_PREFIX}.audit.columnBuilder`,
```

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    const targetDirectoryPath = process.env.PATH_TO_EXPORT_FILES_FOLDER;
+    if (!targetDirectoryPath) {
+      throw new Error('Path to target directory not defined');
+    }
+    data = data.map(e => e.toJSON());
+
+    const targetDirectory = path.join(targetDirectoryPath, '/export-logs/');
+    fs.mkdir(targetDirectory, error =>
+      error
+        ? this.logger.error('Error in Export Logs Directory')
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:31]:
```suggestion
        ? this.logger.error(`An error occurred while creating export logs directory: ${error.toString()}`)
```

Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:31]: If the directory creation is failed, why is execution not being stopped?

Currently it'll just log the message and continue to prepare the excel and try writing that file.

File: services/audit-service/src/services/audit-log-export.service.ts [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+  value(): AuditLogExportFn {
+    return async (data: AnyObject[]) => {
+      return this.auditLogExport(data);
+    };
+  }
+  async auditLogExport(data: AnyObject[]) {
+    const targetDirectoryPath = process.env.PATH_TO_EXPORT_FILES_FOLDER;
+    if (!targetDirectoryPath) {
+      throw new Error('Path to target directory not defined');
```
Comment (shubhamp-sf) [inline services/audit-service/src/services/audit-log-export.service.ts:24]:
```suggestion
      throw new Error('Path to the target directory to export the file is not defined.');
```

File: services/audit-service/src/controllers/export-logs.controller.ts [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+
+      return {jobId: job.getId()};
+    } else {
+      const result = await this.auditLogRepository.find(filter);
+      if (!result) {
+        return {message: 'No data to be exported'};
+      }
+      const customColumnData = await this.columnBuilderService(result);
+      await this.auditLogExportService(customColumnData);
+      return {message: 'File exported successfully'};
```
Comment (shubhamp-sf) [inline services/audit-service/src/controllers/export-logs.controller.ts:83]:
```suggestion
      return {message: 'Audit logs exported successfully.'};
```

**Review by shubhamp-sf — CHANGES_REQUESTED** [review]

File: services/audit-service/src/services/audit-log-export.service.ts (hunk 14 above)
Comment (arpit1503khanna) [reply services/audit-service/src/services/audit-log-export.service.ts:37]: Actually I was handling it in the job processing service , from where this export service is actually called. But I have made the requested change in the exportService

File: services/audit-service/src/component.ts (hunk 1 above)
Comment (arpit1503khanna) [reply services/audit-service/src/component.ts:105]: When I did that , swagger did not displayed controller corresponding to Archive and Export controller

File: services/audit-service/src/services/audit-log-export.service.ts (hunk 14 above)
Comment (shubhamp-sf) [reply services/audit-service/src/services/audit-log-export.service.ts:46]: Handling it there doesn't solves the issue, The edge case you need to handle is if data is present but have no values. Simply an empty array.

Currently you're doing it like this
```ts
if (!data) {
      return;
}
```

Which doesn't solves it as empty array will still be truthy.

File: services/audit-service/src/controllers/export-logs.controller.ts [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+
+  @authenticate(STRATEGY.BEARER)
+  @authorize({
+    permissions: [PermissionKey.ExportLogs, PermissionKey.ExportAuditNum],
+  })
+  @get('/audit-logs/export', {
+    security: OPERATION_SECURITY_SPEC,
+    responses: {
+      [STATUS_CODE.OK]: {
+        description: 'Array of AuditLog model instances',
```
Comment (shubhamp-sf) [inline services/audit-service/src/controllers/export-logs.controller.ts:49]: Wrong description.

**Review by shubhamp-sf — CHANGES_REQUESTED** [review]

File: services/audit-service/src/component.ts (hunk 1 above)
Comment (shubhamp-sf) [reply services/audit-service/src/component.ts:105]: Then try to fix it, there must be some configuration issues.

File: services/audit-service/src/services/audit-log-export.service.ts (hunk 19 above)
Comment (arpit1503khanna) [reply services/audit-service/src/services/audit-log-export.service.ts:43]: done

File: services/audit-service/src/__tests__/fixtures/providers/column-builder.service.ts [hunk 25]
```diff
@@ -0,0 +1,24 @@
+import {Provider} from '@loopback/core';
+import {AuditLog} from '@sourceloop/audit-log';
+import {ColumnBuilderFn} from '../../../types';
+
+// export type ColumnBuilder = unknown;
+
```
[the remark this answers is not in the document]
Comment (arpit1503khanna) [reply services/audit-service/src/__tests__/fixtures/providers/column-builder.service.ts:6]: done

File: services/audit-service/src/services/audit-log-export.service.ts (hunk 14 above)
Comment (arpit1503khanna) [reply services/audit-service/src/services/audit-log-export.service.ts:46]: done

File: services/audit-service/README.md [hunk 26]
```diff
@@ -189,6 +190,16 @@ If `includeArchivedLogs` is set to `true` the data will be fetched from both Aud
 
 If `includeArchivedLogs` option is set to `false` (which is default if not provided) the data is fetched from only Audit Database and not from the archive storage and in this case the response contains the data requested.
 
+### Export-Logs
+
```
[the remark this answers is not in the document]
Comment (arpit1503khanna) [reply services/audit-service/README.md:194]: done

File: services/audit-service/README.md [hunk 27]
```diff
@@ -189,6 +190,16 @@ If `includeArchivedLogs` is set to `true` the data will be fetched from both Aud
...
+This feature is used to export the logs present in Audit Database or the archive storage(eg. AWS.S3). A default loopback filter is accepted based on which logs are exported to the desired location as specified as an excel file(by default). Along with this a boolean variable called `includeArchivedLogs` is also provided which accepts `true` or `false` resulting in whether to include the archieved logs in response.
+
+If `includeArchivedLogs` is set to `true` then data will be fetched from both Audit database and archive storage based on the filter provided as an input **but it is not immediately returned**, a `jobId` is returned which represents the operation happening in the background to fetch and parse logs from archive storage. This `jobId` can be used to check the status of this process and get the result when it is done.
+
+If `includeArchivedLogs` option is set to `false` (which is default if not provided) the data is fetched from only Audit Database and not from the archive storage and in this case the response contains the data requested.
+
+NOTE: The destination path specified for the exported file in the `env` file must exist.
```
[the remark this answers is not in the document]
Comment (arpit1503khanna) [reply services/audit-service/README.md:201]: done

File: services/audit-service/src/__tests__/helpers/db.helper.ts [hunk 28]
```diff
@@ -5,12 +5,20 @@ import {
   AuditLogRepository,
 } from '../../repositories';
 import {archiveLogs} from '../sample-data/archive-log';
-import {ArchiveLogController} from '../../controllers';
-import {JobProcessingService, QuerySelectedFilesProvider} from '../../services';
+import {ArchiveLogController, ExportLogsController} from '../../controllers';
+import {
+  AuditLogExportProvider,
+  // ColumnBuilderProvider,
```
[the remark this answers is not in the document]
Comment (arpit1503khanna) [reply services/audit-service/src/__tests__/helpers/db.helper.ts:11]: done

File: services/audit-service/src/__tests__/unit/export-log.controller.unit.ts [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+} from '../helpers/db.helper';
+import {filterAppliedActedAt} from '../sample-data/filters';
+import {ColumnBuilderProvider} from '../fixtures/providers/column-builder.service';
+import {createStubInstance, expect} from '@loopback/testlab';
+import {AuditLogExportProvider} from '../../services';
+import {Job} from '../../models';
+import {OperationKey} from '../../enums/operation-key.enum';
+import {FileStatusKey} from '../../enums/file-status-key.enum';
+
+describe('POST /audit-logs/archive', () => {
```
[the remark this answers is not in the document]
Comment (arpit1503khanna) [reply services/audit-service/src/__tests__/unit/export-log.controller.unit.ts:14]: done

File: services/audit-service/src/__tests__/helpers/db.helper.ts [hunk 30]
```diff
@@ -5,12 +5,20 @@ import {
...
+  QuerySelectedFilesProvider,
+} from '../../services';
 import {createStubInstance} from '@loopback/testlab';
 import {listMappingLogs, uploaderResponse} from '../sample-data/mapping-log';
 import {AuditServiceApplication} from '../../application';
-import {ExportToCsvFn} from '../../types';
+import {ExportToCsvFn, AuditLogExportFn} from '../../types';
+import {Logger} from '../logger';
+import {ColumnBuilderProvider} from '../fixtures/providers/column-builder.service';
+// import {logger} from 'handlebars';
```
[the remark this answers is not in the document]
Comment (arpit1503khanna) [reply services/audit-service/src/__tests__/helpers/db.helper.ts:21]: done

File: services/audit-service/src/controllers/export-logs.controller.ts (hunk 24 above)
Comment (arpit1503khanna) [reply services/audit-service/src/controllers/export-logs.controller.ts:49]: done

**Review by shubhamp-sf — APPROVED** [review]

**Pull request merged**